pallets/loans-ref/src/lib.rs

```diff
@@ -0,0 +1,792 @@
+				status,
+			});
+
+			Ok(())
+		}
+
+		/// Writes off a loan from admin origin.
+		///
+		/// Forces a writing off of a loan if the `percentage` and `penalty` parameters
+		/// respect the policy values as the minimum.
+		/// This action can write down/up the current write off status of the loan.
+		/// The portfolio valuation of the pool is updated to reflect the new present value of the loan.
+		#[pallet::weight(10_000)]
+		#[transactional]
+		pub fn admin_write_off(
+			origin: OriginFor<T>,
+			pool_id: PoolIdOf<T>,
+			loan_id: T::LoanId,
+			percentage: T::Rate,
+			penalty: T::Rate,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			Self::ensure_role(pool_id, &who, PoolRole::LoanAdmin)?;
+
+			let status = WriteOffStatus {
+				percentage,
+				penalty: Self::to_rate_per_sec(penalty)?,
+			};
+
+			Self::update_active_loan(pool_id, loan_id, |loan| {
+				let state = Self::find_write_off_state(pool_id, loan.maturity_date());
+				let limit = state.map(|s| s.status()).unwrap_or_else(|_| status.clone());
+
+				loan.write_off(&limit, &status)
+			})?;
+
+			Self::deposit_event(Event::<T>::WrittenOff {
+				pool_id,
+				loan_id,
+				status,
+			});
+
+			Ok(())
+		}
+
+		/// Closes a given loan
+		///
+		/// A loan only can be closed if it's fully repaid by the loan borrower.
+		/// Closing a loan gives back the collateral used for the loan to the borrower .
+		#[pallet::weight(10_000)]
+		#[transactional]
+		pub fn close(
+			origin: OriginFor<T>,
+			pool_id: PoolIdOf<T>,
+			loan_id: T::LoanId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let (info, borrower) = match CreatedLoans::<T>::take(pool_id, loan_id) {
+				Some(loan) => (loan.info, loan.borrower),
+				None => Self::take_active_loan(pool_id, loan_id)?.close()?,
+			};
+
+			Self::ensure_loan_borrower(&who, &borrower)?;
+
+			let collateral = *info.collateral();
+			T::NonFungible::transfer(&collateral.0, &collateral.1, &who)?;
+
+			ClosedLoans::<T>::insert(pool_id, loan_id, ClosedLoan::new(info)?);
+
+			Self::deposit_event(Event::<T>::Closed {
+				pool_id,
+				loan_id,
+				collateral,
+			});
+
+			Ok(())
+		}
+
+		/// Updates the write off policy.
+		///
+		/// The write off policy is used to automatically set a write off minimum value to the
+		/// loan.
+		#[pallet::weight(10_000)]
+		#[transactional]
+		pub fn update_write_off_policy(
+			origin: OriginFor<T>,
+			pool_id: PoolIdOf<T>,
+			mut policy: BoundedVec<WriteOffState<T::Rate>, T::MaxWriteOffGroups>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			Self::ensure_role(pool_id, &who, PoolRole::PoolAdmin)?;
+			Self::ensure_pool_exists(pool_id)?;
+
+			policy.iter_mut().try_for_each(|state| -> DispatchResult {
+				state.penalty = Self::to_rate_per_sec(state.penalty)?;
+				Ok(())
+			})?;
+
+			WriteOffPolicy::<T>::insert(pool_id, policy.clone());
+
+			Self::deposit_event(Event::<T>::WriteOffPolicyUpdated { pool_id, policy });
+
+			Ok(())
+		}
```

Comment:
Ok, I see what you mean... 

Should the `write_off()` be called by an automatic process that checks periodically which loans should be written off to avoid this? I thought this was the intended behavior, so "nobody write-off" can not "happen".

Reply:
Yeah it's intended that it's update all at once. If you currently have a policy of 3 days over due, do X, 9 days over due Y, 18 days over due do Z, it doesn't really practically is often relevant to add another group at the end, rather you likely want to change all dates.

> I am not sure, how expensive it is, but can we enforece this logic to iterate over all active loans and update the write-off before updating the given policies?

I don't think this makes much sense, because then immediately after the policy update, the loan can be written off again according to the wnew policy.

Reply:
@offerijns but then write-off policies should go through the `T::ChangeGuard`

Reply:
But agreed, with the above logic, doing what I proposed makes no sense.

Reply:
> @offerijns but then write-off policies should go through the T::ChangeGuard

Ah yes, definitely!

Reply:
I'm not sure exactly how we discussed defining this. From my memory, write-off policy changes should, like pool parameter changes, take 7 days + require no redemptions, since these are significant pool changes that could trigger investors to want to redeem. Whereas loan modifications should only require 1 epoch to pass without redemptions, since these are less impactful changes.

Reply:
Issue here: #1240